### PR TITLE
[SPARK-29551][CORE] Fix a bug about fetch failed when an executor is lost

### PR DESCRIPTION
**Pull request by weixiuli**
### What changes were proposed in this pull request?

When an executor lost with some reason and some things (e.g. the external shuffle service or host lost on the executor's host.) happened, and the executor loses time happens to be reduce stage fetch failed from it which is really bad, the previous only call `mapOutputTracker.unregisterMapOutput(shuffleId, mapIndex, bmAddress) ` to mark clear shuffle status for the mapper at this time , but the other mappers shuffle status on the executor are also unnavailable and the DagScheduler does Not  know that, so the reduce stages will fetch failed again when fetch them,  the unavailable shuffle status can only be resubmitted by a nest retry stage which is the regression.

As we all know that the previous will call ` mapOutputTracker.removeOutputsOnHost(host) `  to  mark clear shuffle status on the host  or `mapOutputTracker.removeOutputsOnExecutor(execId) ` to  mark clear shuffle status on the executor  when reduce stage fetches failed and the executor is active, while it does NOT nothing when the executor lost happened, which  is really bad .

So we should distinguish the failedEpoch of 'executor lost' from the fetchFailedEpoch of 'fetch failed' to solve the above problem.
 
### Why are the changes needed?

The regression  has been described above.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Add unittests.

### CODE REVIEW

**Comment by weixiuli** [comment]
@cloud-fan  @gatorsmile Kindly review, thanks.

**Comment by cloud-fan** [comment]
Can you explain what is the regression? I have no idea what is the problem you are trying to fix.

**Comment by weixiuli** [comment]
When  an executor lost with some reason (eg:. the external shuffle service  or  host lost on the executor's host  ),  while  the executor  loses time happens to be reduce stage  `fetch failed`  from it which is really bad, the previous only call  `mapOutputTracker.unregisterMapOutput(shuffleId, mapIndex, bmAddress)` to mark one map as broken in the map stage  at this time , but other maps on the executor are also not available which can only be resubmitted by a nest retry stage which is the regression.  

So we should distinguish the failedEpoch of 'executor lost' from the fetchFailedEpochof 'fetch failed' to solve the above problem.

As we all know that the previous will call `mapOutputTracker.removeOutputsOnHost(host) `or 
`mapOutputTracker.removeOutputsOnExecutor(execId) ` when reduce stage fetches failed and the executor is active,  while it does NOT for the above problems.

@cloud-fan

**Comment by weixiuli** [comment]
@cloud-fan @gatorsmile @dongjoon-hyun PTAL.

**Comment by tgravescs** [comment]
can you please add the description to the first message. Please change the title to be what is the description of change. Please specify if this is a regression from spark 2.4 or if you are proposing a new feature.  You talk about the specific code in mapoutputtracker but can you please say what the user sees - does the job fail, does it hang, etc.,. fetch failed is a large category.

**Comment by cloud-fan** [comment]
Yea, please describe the problem from an end-user's perspective.

**Comment by gatorsmile** [comment]
cc @jiangxb1987 @Ngone51

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -175,6 +175,10 @@ private[spark] class DAGScheduler(
   // TODO: Garbage collect information about failure epochs when we know there are no more
   //       stray messages to detect.
   private val failedEpoch = new HashMap[String, Long]
+  // There will be a regression when an executor lost and then causes 'fetch failed'.
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: `There will be a regression` is misleading because this mean this PR causes a regression.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 2]
```diff
@@ -537,6 +536,169 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     assert(mapStatus2(2).location.host === "hostB")
   }
 
+  test("All shuffle files on the executor should be cleaned up when executor lost " +
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: For the bug fix PR, we had better add `SPARK-29551` prefix for the test case name.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 3]
```diff
@@ -510,7 +510,6 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     assert(initialMapStatus1.map{_.mapId}.toSet === Set(5, 6, 7))
 
     val initialMapStatus2 = mapOutputTracker.shuffleStatuses(secondShuffleId).mapStatuses
-    //  val initialMapStatus1 = mapOutputTracker.mapStatuses.get(0).get
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: Let's revert this change. This kind clean up is good in general of course, but this makes the backport difficult.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 4]
```diff
@@ -537,6 +536,169 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+          MapStatus(BlockManagerId("exec-hostA1", "hostA", 12345),
+            Array.fill[Long](1)(2), mapTaskId = 8)),
+        (Success,
+          MapStatus(BlockManagerId("exec-hostA2", "hostA", 12345),
+            Array.fill[Long](1)(2), mapTaskId = 9)),
+        (Success, makeMapStatus("hostB", 1))
+      ))
+      // make sure our test setup is correct
+      val initialMapStatus1 = mapOutputTracker.shuffleStatuses(firstShuffleId).mapStatuses
+      //  val initialMapStatus1 = mapOutputTracker.mapStatuses.get(0).get
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: Shall we remove this commented code?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 5]
```diff
@@ -537,6 +536,169 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+      assert(mapStatus1(2).location.host === "hostB")
+
+      val mapStatus2 = mapOutputTracker.shuffleStatuses(secondShuffleId).mapStatuses
+      assert(mapStatus2.count(_ != null) === 2)
+      assert(mapStatus2(0).location.executorId === "exec-hostA1")
+      assert(mapStatus2(0).location.host === "hostA")
+      assert(mapStatus2(2).location.executorId === "exec-hostB")
+      assert(mapStatus2(2).location.host === "hostB")
+    }
+    }
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: Indentation seems to be corrupted in some point before this line.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi @weixiuli .
According to the previous comment, you need to keep the PR description up-to-date.
For example, https://github.com/apache/spark/pull/26206#issuecomment-545234149 should be in the PR description because the PR description will be a commit log eventually.

**Comment by weixiuli** [comment]
@dongjoon-hyun @cloud-fan @tgravescs @jiangxb1987 @Ngone51 @gatorsmile

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 6]
```diff
@@ -537,6 +536,169 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
   }
 
+  test("All shuffle files on the executor should be cleaned up when executor lost " +
+    "and then causes 'fetch failed'") {
+    // whether to kill Executor or not before FetchFailed
+    Seq(true, false).foreach { killExecutor => {
+      afterEach()
+      val conf = new SparkConf()
+      conf.set(config.SHUFFLE_SERVICE_ENABLED.key, "true")
+      conf.set("spark.files.fetchFailure.unRegisterOutputOnHost", "false")
```
[the remark this answers is not in the document]
Comment (weixiuli) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: done.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 4 above)
Comment (weixiuli) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: done.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 7]
```diff
@@ -537,6 +536,169 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+      assert(mapStatus2.count(_ != null) === 2)
+      assert(mapStatus2(0).location.executorId === "exec-hostA1")
+      assert(mapStatus2(0).location.host === "hostA")
+      assert(mapStatus2(2).location.executorId === "exec-hostB")
+      assert(mapStatus2(2).location.host === "hostB")
+    }
+    }
+  }
+
+  test("All shuffle files on the host should be cleaned up when host lost") {
```
[the remark this answers is not in the document]
Comment (weixiuli) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: done.

**Comment by jiangxb1987** [comment]
What problem are you trying to resolve here ？If some jobs failed, can you please attach the related logs in the JIRA ?

**Comment by weixiuli** [comment]
Some more stages of a job will be submitted and the job finish very slowly  without this PR when the above happens . The user-affecting which  I have described  in the  section.

**Comment by joshrosen-stripe** [comment]
Disclaimer: I haven't read this thread in detail yet, so maybe this was already discussed.

Is this PR related to the problems I describe in [SPARK-27736](https://issues.apache.org/jira/browse/SPARK-27736) "Improve handling of FetchFailures caused by ExternalShuffleService losing track of executor registrations"?

**Comment by weixiuli** [comment]
@joshrosen-stripe Yeah, the  PR may solve  the problem which has been  discussed  in   [SPARK-27736](https://issues.apache.org/jira/browse/SPARK-27736).

**Comment by cloud-fan** [comment]
@weixiuli can you describe the problem and your fix at high-level? i.e. do not mention the detailed code.

> As we all know that the previous will callmapOutputTracker.removeOutputsOnHost(host)or
mapOutputTracker.removeOutputsOnExecutor(execId) when reduce stage fetches failed and the executor is active, while it does NOT for the above problems.

This is pretty hard to understand, please use phrases like "clear shuffle status for a host", "clear shuffle status for a mapper", etc.

**Comment by weixiuli** [comment]
@cloud-fan @dongjoon-hyun @tgravescs @jiangxb1987 PTAL.
